### PR TITLE
feat: add support for proper parsing to dracula

### PR DESCRIPTION
**Pull request by swarnim-deepsource**
use tree-sitter to provide proper means of parsing.

And add support for several languages in the v2 function to the pydracula library.

Tasks:

* [x] Add `v2` parsing using tree-sitter for more accurate analysis.
* [x] Add support for all languages required or possibly required at DeepSource.
* [x] Adapt old tests to verify the new changes.
* [x] Add further tests for newly supported languages(we have only exposed js and typescript for now via python bindings). 

Fixes lae-7716

### CODE REVIEW

**Comment by tushar-deepsource** [comment]
@swarnim-deepsource is V2 more reliable than V1? Should we switch test coverage completely to v2?

File: src/langs/mod.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+mod rust;
+mod scala;
+
+pub use java::Java;
+pub use c::C;
+pub use csharp::Csharp;
+pub use python::Python;
+pub use ruby::Ruby;
+pub use rust::Rust;
+pub use scala::Scala;
```
Comment (prajwal-deepsource) [inline src/langs/mod.rs:17]: Not exposing js and jsx?

Comment (swarnim-deepsource) [reply src/langs/mod.rs:17]: that's v1, look into pydracula.

Comment (prajwal-deepsource) [reply src/langs/mod.rs:17]: Ack, got it

**Review by prajwal-deepsource — DISMISSED** [review]

File: src/tests.rs [hunk 2]
```diff
@@ -348,10 +399,118 @@ mod simple_rust {
...
+
+    #[test]
+    fn try_parse() {
+        let src = r#"
+        function test(_: any) { // :01
+            // empty
+            let x: // :02
+                Ty = 10; // :03
+            let y = ` // :04
+                /* this is part of the string? */ // :04 as this is template string
```
Comment (tushar-deepsource) [inline src/tests.rs:500]: why `:04`? should be `:05` right

**Review by tushar-deepsource — COMMENTED** [review]
I'd appreciate it if you can have tests asserting the cleaned source as well, not just line count. Will be more robust.

File: src/tests.rs (hunk 2 above)
Comment (swarnim-deepsource) [reply src/tests.rs:500]: typo :P

Comment (swarnim-deepsource) [inline src/tests.rs:500]:
```suggestion
                /* this is part of the string? */ // :05 as this is template string
```

**Comment by swarnim-deepsource** [comment]
@tushar-deepsource we don't generate clean_src any more. We don't need to worry about cleaned source because the parsing is out of our hands now, and can be assumed to correct unless it fails cleanly.
I'd appreciate it if you can have tests asserting the cleaned source as well, not just line count. Will be more robust.

**Comment by tushar-deepsource** [comment]
> and can be assumed to correct unless it fails cleanly.
No I disagree, even if the code is correct the test is not good enough. By just checking the number of valid lines == 7 can pass even if the code is wrong. So at least asserting the exact valid indices would be needed, not just their count.

**Comment by swarnim-deepsource** [comment]
> No I disagree, even if the code is correct the test is not good enough. By just checking the number of valid lines == 7 can pass even if the code is wrong. So at least asserting the exact valid indices would be needed, not just their count.

@tushar-deepsource we already do index matching, in the run_on_self function.

File: src/tests.rs [hunk 3]
```diff
@@ -46,15 +63,29 @@ mod simple_python {
...
+                """
                 Multi-line Comments
                 """
                 print(x)
                 """
                 Multi-line Comments
                 """
             "#;
-        assert_eq!(get_count_of_meaningful_lines::<Python>(src), 5)
+        // let lines = count::get_meaningful_line_indices::<Python>(src)
```
Comment (tushar-deepsource) [inline src/tests.rs:75]: turn these into assertions

File: src/tests.rs [hunk 4]
```diff
@@ -46,15 +63,29 @@ mod simple_python {
...
+        let cnt = count::get_count_of_meaningful_lines::<Python>(src);
+        let cnt_executable = {
+            let mut parser = Parser::new(TreeSitterLanguage::Python).unwrap();
+            let ranges = parser.non_executable_src_spans(src).unwrap();
+            let lines = get_lines_without_ranges(src, ranges);
+            // crate::parse::v2::display_lines(src, &lines);
+            lines.len()
+        };
+        assert_eq!(cnt, 5);
+        assert_eq!(cnt_executable, 5);
```
Comment (tushar-deepsource) [inline src/tests.rs:88]: assert indices as well, with the total count

**Review by tushar-deepsource — CHANGES_REQUESTED** [review]

**Review by tushar-deepsource — APPROVED** [review]

**Pull request merged**